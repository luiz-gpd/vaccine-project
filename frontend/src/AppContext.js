import { createContext } from 'react'

const AppContext = createContext();


// const reducer = (state, action) => {
//     switch (action.type) {
//         case "EXEMPLE": {
//             return {
//                 ...state,
//                 loggedUser: action.payload,
//             };
//         }

//         default: {
//             return state;
//         }
//     }
// }

export default AppContext