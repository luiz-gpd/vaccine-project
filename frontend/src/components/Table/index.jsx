import React, { useEffect, useContext } from 'react'
import moment from 'moment'
import api from '../../utils/api'
import { ClayToggle } from '@clayui/form'
import { ClayButtonWithIcon } from '@clayui/button'
import AppContext from '../../AppContext'

const Table = ( { usersPerPage, pageNumber } ) => {

    const [{ search }, dispatch] = useContext(AppContext);
    const [users, setUsers] = React.useState([])

    const getUsers = async () => {
        api.get('/user').then((response) => {
            const { data } = response;
            setUsers(data);
        }).catch((e) => {
            console.log(e);
        });
    };

    useEffect(() => {
        getUsers();
    }, [])

    const onClick = (id, info) => {
        dispatch({
            type: "SET_MODAL",
            payload: {
                userId: id,
                info: info,
                visible: true,
            },
          });
    }

    const onToggle = async (id, attended) => {
        const formData = {
            attended: !attended
        }
        await api.put(`/user/${id}`, formData);
        getUsers();
    }

    let pagesVisited = pageNumber * usersPerPage;

    return (
        <table className="content-table">
                        <thead>
                            <tr>
                                <th>Nome</th>
                                <th>Idade</th>
                                <th>Data da Vacina</th>
                                <th>Horário da Vacina</th>
                                <th>Atendimento</th>
                            </tr>
                        </thead>
                        <tbody>
                            {users.length && (
                            // eslint-disable-next-line array-callback-return
                            users.filter((user)=> {
                                    if (search === "") {
                                        return user;
                                    } else if(user.name.toLowerCase().includes(search.toLowerCase())) {
                                        pagesVisited=10;
                                        return user;
                                    }
                                })
                                .sort((
                                    (a, b) =>
                                        (moment(a.consultationDate).isBefore(b.consultationDate)) ? 
                                            moment(a.consultationDate).diff(b.consultationDate) :
                                            (a.consultationDate === b.consultationDate) &&
                                            (moment(moment(a.consultationTime, "HH:mm")).diff(moment(b.consultationTime, "HH:mm")))
                                ))
                                .slice(pagesVisited - usersPerPage, pagesVisited)
                                .map((user, key) => (
                                    <tr key={key}>
                                        <td>{user.name}</td>
                                        <td>{user.age}</td>
                                        <td>{moment(user.consultationDate).format("DD/MM/yyyy")}</td>
                                        <td>{user.consultationTime}</td>
                                        <td><ClayToggle label={user.attended ? "Realizado" : "Não foi realizado"}
                                                disabled={(moment(user.consultationDate).isBefore(new Date())) ? false : true}
                                                toggled={user.attended}
                                                data-testid='table-toggle'
                                                onToggle={() => onToggle(user._id, user.attended)}/>
                                            {user.attended && <ClayButtonWithIcon data-testid="table-button" className="btn btn-primary btn-sm ml-2"
                                                symbol="comments" onClick={() => onClick(user._id, user.consultInfo)} />}</td>
                                    </tr>

                                )))}
                        </tbody>
                    </table>
    )
}

export default Table;