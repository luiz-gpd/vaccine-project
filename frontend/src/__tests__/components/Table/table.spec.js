import Table from "../../../components/Table";
import React from 'react'
import AppContext, { initialState } from "../../../AppContext";
import { render } from "@testing-library/react";

const TableWrapper = ({ state = initialState, dispatch }) => (
  <AppContext.Provider value={[state, dispatch]}>
    <Table />
  </AppContext.Provider>
);

describe("Table Component", () => {
  it("renders", () => {
    const { asFragment } = render(<TableWrapper />);

    expect(asFragment()).toMatchSnapshot();
  });
  it("renders with context when search input is empty, and dates are different", () => {
    const dispatch = jest.fn();

    const myInitialState =
      [{
        name: "Luiz",
        age: "19",
        consultationDate: "2021-04-30T03:00:00.000Z",
        consultationTime: "08:30",
        attended: false,
        consultInfo: ""
      },
      {
        name: "Gustavo",
        age: "20",
        consultationDate: "2021-04-29T03:00:00.000Z",
        consultationTime: "09:30",
        attended: false,
        consultInfo: ""
      },
      ]

    React.useState = jest.fn().mockReturnValue([myInitialState, {}])

    const { queryByText } = render(
      <TableWrapper
        state={{
          ...initialState,
          search: ""
        }}
        usersPerPage="2"
        pageNumber="1"
        dispatch={dispatch}
      />
    );

    expect(queryByText("Nome")).toBeTruthy();
    expect(queryByText("Idade")).toBeTruthy();
    expect(queryByText("HorÃ¡rio da Vacina")).toBeTruthy();
    expect(queryByText("Atendimento")).toBeTruthy();

  });
  it("renders with context when search input is full, and dates are the same",() => {
      const dispatch = jest.fn();

      const myInitialState =
        [{
          name: "Luiz",
          age: "19",
          consultationDate: "2021-04-30T03:00:00.000Z",
          consultationTime: "08:30",
          attended: false,
          consultInfo: ""
        },
        {
          name: "Gustavo",
          age: "20",
          consultationDate: "2021-04-30T03:00:00.000Z",
          consultationTime: "09:30",
          attended: true,
          consultInfo: ""
        },
        ]

      React.useState = jest.fn().mockReturnValue([myInitialState, {}])

      render(
        <TableWrapper
          state={{
            ...initialState,
            search: "u"
          }}
          usersPerPage="10"
          pageNumber="1"
          dispatch={dispatch}
        />
      );
    });
});