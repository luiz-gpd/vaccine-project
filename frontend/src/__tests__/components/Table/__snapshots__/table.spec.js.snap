// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component renders 1`] = `
<DocumentFragment>
  <table
    class="content-table"
  >
    <thead>
      <tr>
        <th>
          Nome
        </th>
        <th>
          Idade
        </th>
        <th>
          Data da Vacina
        </th>
        <th>
          Hor√°rio da Vacina
        </th>
        <th>
          Atendimento
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  0
</DocumentFragment>
`;
